### ISSUE TITLE
Richmondshire council collection schedule - XML data

### ISSUE THREAD

**Issue by theanorak-keith** [mampfes/hacs_waste_collection_schedule#329]
Richmondshire council in Yorkshire, England has a data API for their waste collection schedule here:
https://www.richmondshire.gov.uk/Umbraco/Api/MyAreaApi/GetBinRoundData?uprn=

On their website, they have a postcode search which populates a dropdown list which uses UPRNs as the 'value' field — I'd say it's reasonable to ask people to find their UPRN :)

The API outputs XML in this format:
```
<ArrayOfBinRoundEntity xmlns:i="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://schemas.datacontract.org/2004/07/Richmondshire.controllers">
<BinRoundEntity>
<end>2022-08-12T00:00:00</end>
<id>114991</id>
<start>2022-08-12T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-08-19T00:00:00</end>
<id>34853</id>
<start>2022-08-19T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-08-26T00:00:00</end>
<id>114991</id>
<start>2022-08-26T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-09-03T00:00:00</end>
<id>34853</id>
<start>2022-09-03T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-09-09T00:00:00</end>
<id>114991</id>
<start>2022-09-09T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-09-16T00:00:00</end>
<id>34853</id>
<start>2022-09-16T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-09-23T00:00:00</end>
<id>114991</id>
<start>2022-09-23T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-09-30T00:00:00</end>
<id>34853</id>
<start>2022-09-30T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-10-07T00:00:00</end>
<id>114991</id>
<start>2022-10-07T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-10-14T00:00:00</end>
<id>34853</id>
<start>2022-10-14T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-10-21T00:00:00</end>
<id>114991</id>
<start>2022-10-21T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-10-28T00:00:00</end>
<id>34853</id>
<start>2022-10-28T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-11-04T00:00:00</end>
<id>114991</id>
<start>2022-11-04T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-11-11T00:00:00</end>
<id>34853</id>
<start>2022-11-11T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-11-18T00:00:00</end>
<id>114991</id>
<start>2022-11-18T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-11-25T00:00:00</end>
<id>34853</id>
<start>2022-11-25T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-12-02T00:00:00</end>
<id>114991</id>
<start>2022-12-02T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-12-09T00:00:00</end>
<id>34853</id>
<start>2022-12-09T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-12-16T00:00:00</end>
<id>114991</id>
<start>2022-12-16T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-12-23T00:00:00</end>
<id>34853</id>
<start>2022-12-23T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2022-12-30T00:00:00</end>
<id>114991</id>
<start>2022-12-30T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-01-06T00:00:00</end>
<id>34853</id>
<start>2023-01-06T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-01-13T00:00:00</end>
<id>114991</id>
<start>2023-01-13T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-01-20T00:00:00</end>
<id>34853</id>
<start>2023-01-20T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-01-27T00:00:00</end>
<id>114991</id>
<start>2023-01-27T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-02-03T00:00:00</end>
<id>34853</id>
<start>2023-02-03T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-02-10T00:00:00</end>
<id>114991</id>
<start>2023-02-10T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-02-17T00:00:00</end>
<id>34853</id>
<start>2023-02-17T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-02-24T00:00:00</end>
<id>114991</id>
<start>2023-02-24T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-03-03T00:00:00</end>
<id>34853</id>
<start>2023-03-03T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-03-10T00:00:00</end>
<id>114991</id>
<start>2023-03-10T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-03-17T00:00:00</end>
<id>34853</id>
<start>2023-03-17T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-03-24T00:00:00</end>
<id>114991</id>
<start>2023-03-24T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-03-31T00:00:00</end>
<id>34853</id>
<start>2023-03-31T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-04-07T00:00:00</end>
<id>114991</id>
<start>2023-04-07T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-04-14T00:00:00</end>
<id>34853</id>
<start>2023-04-14T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-04-21T00:00:00</end>
<id>114991</id>
<start>2023-04-21T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-04-28T00:00:00</end>
<id>34853</id>
<start>2023-04-28T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-05-05T00:00:00</end>
<id>114991</id>
<start>2023-05-05T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-05-12T00:00:00</end>
<id>34853</id>
<start>2023-05-12T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-05-19T00:00:00</end>
<id>114991</id>
<start>2023-05-19T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-05-26T00:00:00</end>
<id>34853</id>
<start>2023-05-26T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-06-02T00:00:00</end>
<id>114991</id>
<start>2023-06-02T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-06-09T00:00:00</end>
<id>34853</id>
<start>2023-06-09T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-06-16T00:00:00</end>
<id>114991</id>
<start>2023-06-16T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-06-23T00:00:00</end>
<id>34853</id>
<start>2023-06-23T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-06-30T00:00:00</end>
<id>114991</id>
<start>2023-06-30T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-07-07T00:00:00</end>
<id>34853</id>
<start>2023-07-07T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-07-14T00:00:00</end>
<id>114991</id>
<start>2023-07-14T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-07-21T00:00:00</end>
<id>34853</id>
<start>2023-07-21T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-07-28T00:00:00</end>
<id>114991</id>
<start>2023-07-28T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-08-04T00:00:00</end>
<id>34853</id>
<start>2023-08-04T00:00:00</start>
<title>55L RECYCLING BOX</title>
</BinRoundEntity>
<BinRoundEntity>
<end>2023-08-11T00:00:00</end>
<id>114991</id>
<start>2023-08-11T00:00:00</start>
<title>240L GREY RUBBISH BIN</title>
</BinRoundEntity>
</ArrayOfBinRoundEntity>
```

Are there any existing integrations that parse XML that I could use as a jumping-off point? Most of the ones I opened at random seemed to be expecting JSON data, and rewriting one of those to parse XML is likely beyond my skills :(

**Comment by dt215git** [comment]
There a several ways of parsing xml in python.
A lot of the scripts here use BeautifulSoup to parse html, but it can also do xml, so perhaps try with that? You might need an xml parser alongside bs4, something like `pip install lxml`

As an example, with your xml response saved into a _testxml.xml_ file

```
from bs4 import BeautifulSoup

with open('testxml.xml') as f:
    testxml = f.read()
f.close()

soup = BeautifulSoup(testxml, "xml")
dates = soup.find_all('start')
rounds = soup.find_all('title')

for i in range(0, len(dates)):
    print(dates[i].text, ':', rounds[i].text)
```

Gives:

```
2022-08-12T00:00:00 : 240L GREY RUBBISH BIN
2022-08-19T00:00:00 : 55L RECYCLING BOX
2022-08-26T00:00:00 : 240L GREY RUBBISH BIN
2022-09-03T00:00:00 : 55L RECYCLING BOX
..... lots more lines .....
2023-07-21T00:00:00 : 55L RECYCLING BOX
2023-07-28T00:00:00 : 240L GREY RUBBISH BIN
2023-08-04T00:00:00 : 55L RECYCLING BOX
2023-08-11T00:00:00 : 240L GREY RUBBISH BIN
```

**Comment by theanorak-keith** [comment]
Thank you for this.

I tried this tonight, and all was looking promising, but I kept getting 'found 0 entries' so eventually I added a 'print' statement to my new source file, so I could check it was receiving data and not a 'go away' message...and discovered that the API *is* sending JSON data, but Chrome helpfully formatted it as XML and I didn't think to check.

So I feel a bit foolish now, but at least I learned something.
